### PR TITLE
Polymer 1.0 demo cleanup

### PR DESCRIPTION
**Pull request by ebidel**
R: @addyosmani 

- Polymer 1.1.4
- Use lite wc.js polyfill and conditionally load iff they're needed.
- element dep updates
- remove hacks
- clean up element logic and use more polymer features and idiomatic patterns
- use vulcanized html import bundle with `async`.

### CODE REVIEW

**Comment by samccone** [comment]
you will want to add ../examples/polymer/elements/elements.build.js to out .jscs ignore

**Comment by samccone** [comment]
woohoo tests passing :clap:

**Comment by ebidel** [comment]
Merge?

**Comment by samccone** [comment]
going to let @arthurvr or @addyosmani also take a peak

File: examples/polymer/elements/td-todos.html [hunk 1]
```diff
@@ -6,140 +6,142 @@
 
 <dom-module id="td-todos">
 	<template>
-		<flatiron-director id="router" route="{{route}}"></flatiron-director>
+	  <style>
```
Comment (arthurvr) [inline examples/polymer/elements/td-todos.html]: Wrong indentation :)

File: examples/polymer/elements/td-model.html [hunk 2]
```diff
@@ -27,50 +38,39 @@
...
-					this.items.forEach(function(item, i) {
-						if (this.filter) {
-							if (this.filters[this.filter](item)) {
-								this.splice('items', i, 1, {title: item.title, completed: completed});
-							}
-						} else {
-							this.splice('items', i, 1, {title: item.title, completed: completed});
-						}
-					}, this);
+          for (var i = 0; i < this.items.length; ++i) {
```
Comment (arthurvr) [inline examples/polymer/elements/td-model.html]: Wrong indentation :)

File: examples/polymer/elements/td-item.html [hunk 3]
```diff
@@ -28,49 +35,48 @@
 					},
 					item: {
 						type: Object,
-						value: function() {
-							return {};
-						}
+						value: function() { return {}; }
```
Comment (arthurvr) [inline examples/polymer/elements/td-item.html]: On this and quite a bit of other places, `function () {}` (spacing)

File: examples/polymer/package.json [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  "keywords": [
+    "polymer",
+    "todomvc"
+  ],
+  "author": "",
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/tastejs/todomvc/issues"
+  },
+  "homepage": "https://github.com/tastejs/todomvc",
```
Comment (arthurvr) [inline examples/polymer/package.json]: Remove everything except the `devDependencies` and add `private` instead :)

File: examples/polymer/index.html [hunk 5]
```diff
@@ -3,28 +3,41 @@
...
+			    && 'import' in document.createElement('link')
+			    && 'content' in document.createElement('template'));
+
+				if (!webComponentsSupported) {
+				  var script = document.createElement('script');
+				  script.async = true;
+				  script.src = 'bower_components/webcomponentsjs/webcomponents-lite.min.js';
+				  document.head.appendChild(script);
+				}
+			})();
```
Comment (arthurvr) [inline examples/polymer/index.html]: Wrong indentation (this whole `script` tag)

File: examples/polymer/index.html [hunk 6]
```diff
@@ -3,28 +3,41 @@
...
 		<meta charset="utf-8">
 		<title>Polymer • TodoMVC</title>
-		<script src="bower_components/webcomponentsjs/webcomponents.js"></script>
 		<link rel="stylesheet" href="bower_components/todomvc-common/base.css">
 		<link rel="stylesheet" href="bower_components/todomvc-app-css/index.css">
-		<link rel="import" href="bower_components/iron-localstorage/iron-localstorage.html">
-		<link rel="import" href="elements/td-model.html">
-		<link rel="import" href="elements/td-todos.html">
+		<link rel="import" href="elements/elements.build.html" async>
+		<!-- <link rel="import" href="elements/elements.html"> for dev -->
```
Comment (arthurvr) [inline examples/polymer/index.html]: Might want to add a note about this to the readme.

File: examples/polymer/elements/td-todos.html [hunk 7]
```diff
@@ -6,140 +6,142 @@
...
+				_computeItemsLeft: function(count) {
+					return count < 2 ? 'item' : 'items';
 				},
-				routeChanged: function(e) {
-					this.model.filter = e.detail;
+
+				matchesFilter: function(route) {
+					return function(item, index, array) {
+						return this.model ? this.model.matchesFilter(item, route) : false;
+				  }.bind(this);
```
Comment (arthurvr) [inline examples/polymer/elements/td-todos.html]: Indentation :)

**Comment by arthurvr** [comment]
Generally our policy would be not to have a build system (even if it's as easy as `vulcanize`). I don't see why this example would need to be an exception.

File: examples/polymer/vulcanize.sh [hunk 8]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+
+# vulcanize elements.html \
+#     --inline-script --inline-css --strip-comments | \
+#     crisper -h elements.build.html -j elements.build.js
+
+polybuild --maximum-crush elements/elements.html
```
Comment (arthurvr) [inline examples/polymer/vulcanize.sh]: And if we actually are going to add this, please use a `package.json` script. No reason to have a file like this.

File: examples/polymer/index.html (hunk 5 above)
Comment (arthurvr) [inline examples/polymer/index.html]: Install EditorConfig and you wouldn't have to deal with that :)

File: examples/polymer/vulcanize.sh (hunk 8 above)
Comment (addyosmani) [inline examples/polymer/vulcanize.sh]: I think we should drop `vulcanize.sh` and just move this into an `npm run` script instead. It's a one liner that would fit well in package.json.

**Comment by addyosmani** [comment]
Reviewed. On the whole this looks solid. @ebidel if you could address the review comments, we should be able to get this landed shortly.

File: examples/polymer/package.json (hunk 4 above)
Comment (ebidel) [inline examples/polymer/package.json]: Done

File: examples/polymer/index.html (hunk 5 above)
Comment (ebidel) [inline examples/polymer/index.html]: Who uses tabs!

File: examples/polymer/elements/td-model.html (hunk 2 above)
Comment (ebidel) [inline examples/polymer/elements/td-model.html]: Done

File: examples/polymer/elements/td-item.html (hunk 3 above)
Comment (ebidel) [inline examples/polymer/elements/td-item.html]: Done. Weird that anon functions are treated differently than named functions.

File: examples/polymer/vulcanize.sh (hunk 8 above)
Comment (arthurvr) [inline examples/polymer/vulcanize.sh]: > I think one of the goals of this update was to introduce Polymer performance patterns which require a build step in this case.

Same can be said about the other frameworks build systems, which we never allowed. I do a agree that the apps should present best practices but TodoMVC isn't around performance. It's way more important that people can read the code easily and contribute back without having to setup too much. The minified elements.html for example, is against what we decided in https://github.com/tastejs/todomvc/issues/378.

Comment (arthurvr) [inline examples/polymer/vulcanize.sh]: cc @tastejs/todomvc for input from the rest of the team.

File: examples/polymer/elements/td-item.html (hunk 3 above)
Comment (arthurvr) [inline examples/polymer/elements/td-item.html]: They aren't treated differently? There's simply a space after every keyword (`function`, `while`, `if`, ..) It's inconsistent not to have one after function :)

File: examples/polymer/index.html (hunk 6 above)
Comment (ebidel) [inline examples/polymer/index.html]: Done

File: examples/polymer/elements/td-todos.html (hunk 7 above)
Comment (ebidel) [inline examples/polymer/elements/td-todos.html]: Done

File: examples/polymer/vulcanize.sh (hunk 8 above)
Comment (ebidel) [inline examples/polymer/vulcanize.sh]: Removed

We need to keep vulcanize for a couple of reasons:

1. We encourage developers to use this tool for performance as a best practice. Our canonical todo app should be doing the same.
1. Polymer is different than other frameworks b/c it requires polyfills in other browsers. This app isnt too large, but html imports not being native in most browsers means slower perf in those browsers (if you're not vulcanizing). The last thing we want is devs running the demo and coming up with the the misperception that "Polymer is slow" b/c of this hinderance.

File: examples/polymer/README.md [hunk 9]
```diff
@@ -42,3 +41,11 @@ Polymer and its polyfills are intended to work in the latest version of [evergre
      `python -m SimpleHTTPServer`
 
 1. Browse to the server root
+
+## Making updates
+
+If you want to make a change to any of the elements in elements/elements.html, you'll need to install `polybuild` (Polymer's build tool) and re-vulcanize elements.build.html.
+
+1. Run `npm install`
+1. Make a change
```
Comment (arthurvr) [inline examples/polymer/README.md]: This ordering seems wrong :)

File: examples/polymer/index.html [hunk 10]
```diff
@@ -3,28 +3,41 @@
...
 		<footer id="info">
 			<p>Double-click to edit a todo</p>
-			<p>Created by <a href="http://www.polymer-project.org">The Polymer Authors</a></p>
+			<p>Created by <a href="https://www.polymer-project.org/1.0/">The Polymer Authors</a></p>
 			<p>Part of <a href="http://todomvc.com">TodoMVC</a></p>
 		</footer>
-		<script src="bower_components/todomvc-common/base.js"></script>
+		<script src="bower_components/todomvc-common/base.js" async></script>
+		<script>
+			(function() {
```
Comment (arthurvr) [inline examples/polymer/index.html]: `function () {`

File: examples/polymer/README.md (hunk 9 above)
Comment (ebidel) [inline examples/polymer/README.md]: Seems right to me. You only need the tool if you're making changes. Need to run install, make a change, then rebuild.

Comment (arthurvr) [inline examples/polymer/README.md]: I meant that you had `1.` twice :)

File: examples/polymer/index.html (hunk 10 above)
Comment (ebidel) [inline examples/polymer/index.html]: Done

**Comment by ebidel** [comment]
PTAL

**Comment by arthurvr** [comment]
After https://github.com/tastejs/todomvc/pull/1455#discussion_r41088371 is resolved this LGTM.

File: examples/polymer/README.md (hunk 9 above)
Comment (ebidel) [inline examples/polymer/README.md]: I was following the existing readme examples, but github will render a sequence: https://github.com/tastejs/todomvc/tree/gh-pages/examples/polymer#running-this-sample.

**Comment by ebidel** [comment]
PTAL

File: examples/polymer/README.md (hunk 9 above)
Comment (arthurvr) [inline examples/polymer/README.md]: I know, but then be consistent and eighter use `1.` for all of them or use the right order. Now the last item uses `2.`...

Comment (ebidel) [inline examples/polymer/README.md]: Sorry, missed that one. I eat my own words on consistency!

**Comment by ebidel** [comment]
How we lookin?

**Comment by addyosmani** [comment]
This LGTM to me for a land now. Thanks for working through the review, @ebidel!

**Comment by arthurvr** [comment]
Thanks! Nice work! :100: 

Can we make sure to squash the commits next time? :)

**Pull request merged**